src/main/kotlin/com/terraformation/backend/nursery/api/BatchesHistoryController.kt

```diff
@@ -0,0 +1,339 @@
+    override val version: Int,
+) : BatchHistoryPayload {
+  constructor(
+      row: BatchDetailsHistoryRow,
+      subLocationRows: Collection<BatchDetailsHistorySubLocationsRow>?
+  ) : this(
+      createdBy = row.createdBy!!,
+      createdTime = row.createdTime!!,
+      notes = row.notes,
+      projectId = row.projectId,
+      projectName = row.projectName,
+      readyByDate = row.readyByDate,
+      subLocations = subLocationRows?.map { BatchHistorySubLocationPayload(it) } ?: emptyList(),
+      substrate = row.substrateId,
+      substrateNotes = row.substrateNotes,
+      treatment = row.treatmentId,
+      treatmentNotes = row.treatmentNotes,
+      version = row.version!!,
+  )
+
+  override val type: BatchHistoryPayloadType
+    get() = BatchHistoryPayloadType.DetailsEdited
+}
+
+@Schema(description = "A manual edit of a batch's remaining quantities.")
+data class BatchHistoryQuantityEditedPayload(
+    override val createdBy: UserId,
+    override val createdTime: Instant,
+    val germinatingQuantity: Int,
+    val notReadyQuantity: Int,
+    val readyQuantity: Int,
+    override val version: Int,
+) : BatchHistoryPayload {
+  constructor(
+      row: BatchQuantityHistoryRow
+  ) : this(
+      createdBy = row.createdBy!!,
+      createdTime = row.createdTime!!,
+      germinatingQuantity = row.germinatingQuantity!!,
+      notReadyQuantity = row.notReadyQuantity!!,
+      readyQuantity = row.readyQuantity!!,
+      version = row.version!!,
+  )
+
+  override val type: BatchHistoryPayloadType
+    get() = BatchHistoryPayloadType.QuantityEdited
+}
+
+@Schema(
+    description =
+        "The new quantities resulting from changing the statuses of seedlings in a batch. The " +
+            "values here are the total quantities remaining after the status change, not the " +
+            "number of seedlings whose statuses were changed.")
+data class BatchHistoryStatusChangedPayload(
+    override val createdBy: UserId,
+    override val createdTime: Instant,
+    val germinatingQuantity: Int,
+    val notReadyQuantity: Int,
+    val readyQuantity: Int,
+    override val version: Int,
+) : BatchHistoryPayload {
+  constructor(
+      row: BatchQuantityHistoryRow
+  ) : this(
+      createdBy = row.createdBy!!,
+      createdTime = row.createdTime!!,
+      germinatingQuantity = row.germinatingQuantity!!,
+      notReadyQuantity = row.notReadyQuantity!!,
+      readyQuantity = row.readyQuantity!!,
+      version = row.version!!,
+  )
+
+  override val type: BatchHistoryPayloadType
+    get() = BatchHistoryPayloadType.StatusChanged
+}
+
+@Schema(
+    description =
+        "A nursery transfer withdrawal from another batch that added seedlings to this batch.")
+data class BatchHistoryIncomingWithdrawalPayload(
+    override val createdBy: UserId,
+    override val createdTime: Instant,
+    val fromBatchId: BatchId,
+    val germinatingQuantityAdded: Int,
+    val notReadyQuantityAdded: Int,
+    val readyQuantityAdded: Int,
```

Comment:
Feedback wanted on this!

These numbers are deltas, unlike the quantities in the "edited" and "status changed" payloads which are the absolute values. This might not be the right design decision; I can easily make this return the absolute values instead. I did it this way because this matches the way the numbers work in the withdrawal history.

Reply:
I think it makes sense like this, and it seems consistent from a naming perspective so it should be relatively easy to understand.